trl/extras/remote_vllm_helper.py

```diff
@@ -0,0 +1,401 @@
+"""
+Easy vllm with nccl link with training process, build for LLM researchers
+
+@Binary-Husky
+
+launch vllm fastapi server (infer + nccl link)
+    ```
+    REMOTE_VLLM_NCCL_LINK=True \
+    REMOTE_VLLM_GPUS=4 \
+    CUDA_VISIBLE_DEVICES="0,1,2,3" \
+    REMOTE_VLLM_INIT_MODEL=/mnt/data_cpfs/model_cache/modelscope/hub/Qwen/Qwen/QwQ-32B \
+    REMOTE_VLLM_GPU_FRAG=0.95 \
+    REMOTE_VLLM_MAX_MODEL_LEN=8192 \
+    REMOTE_VLLM_MAX_LORA_RANK=0 \
+    REMOTE_VLLM_TEMPERATURE=0.9 \
+    REMOTE_VLLM_NUM_GENERATION=1 \
+    python vllm_ops.py
+    ```
+
+client demo (infer + nccl link)
+    ```
+    from trl.extras.remote_vllm_helper import VllmRemoteClient
+    self.remote_llm = VllmRemoteClient(nccl_link=True)
+    self.remote_llm.update_attr(key="temperature", value=args.temperature)
+    self.remote_llm.update_attr(key="max_completion_length", value=self.max_completion_length)
+    self.remote_llm.update_attr(key="vllm_max_model_len", value=self.args.vllm_max_model_len)
+    self.remote_llm.update_attr(key="num_generations", value=args.num_generations)
+
+    # read zero3 param and update model [assert not is_peft_model(self.model)]
+    import deepspeed
+    import time
+    import tqdm
+    for name, param in tqdm.tqdm(self.model.named_parameters(), total=len(list(self.model.named_parameters()))):
+        with deepspeed.zero.GatheredParameters([param], enabled=self.accelerator.state.deepspeed_plugin.zero_stage == 3):
+            if self.accelerator.is_main_process:
+                weight = param.data
+                assert weight.shape == param.ds_shape
+                self.remote_llm.advanced_update_param(name, weight)
+                logger.error(str(weight.sum().item()) + "\t" +  str(param.shape) + "\t" + str(param.ds_shape))
+    import gc
+    gc.collect()
+    torch.cuda.empty_cache()
+
+    # predict as usual
+    prompt = tokenizer.apply_chat_template(conversation=[
+        {"role": "user", "content": "黄金的延展性"},
+    ], tokenize=False, add_generation_prompt=True)
+    tokens = vrc.generate([prompt])
+    result = tokenizer.batch_decode(tokens)
+    ```
+
+"""
+
+
+import os
+import sys
+import time
+import torch
+import uvicorn
+import requests
+import threading
+import cloudpickle as pickle
+
+# from loguru import logger
+import logging
+logger = logging.getLogger(__name__)
+
+from vllm.worker.worker import Worker
+from vllm.lora.request import LoRARequest
+from fastapi import FastAPI, Request, Response, HTTPException
+
+os.environ["VLLM_WORKER_MULTIPROC_METHOD"] = "spawn"
+
+ENV = {
```

Comment:
Where possible, it would be good to argparse as many of these options, rather than relying on env vars